File: mobile/ios/packages/town-crier-domain/Sources/Entities/EntitlementMap.swift
/// Maps subscription tiers to their granted entitlements and quota limits.
///
/// Must remain in sync with the API's `EntitlementMap.cs`.
/// This is a caseless enum because it has no instances — all access is through static methods.
public enum EntitlementMap {

  // MARK: - Entitlement sets per tier

  private static let paidEntitlements: Set<Entitlement> = [
    .statusChangeAlerts,
    .decisionUpdateAlerts,
    .hourlyDigestEmails,
  ]

  // MARK: - Public API

  /// Returns the set of entitlements granted to the given subscription tier.
  public static func entitlements(for tier: SubscriptionTier) -> Set<Entitlement> {
    switch tier {
    case .free:
      return []
    case .personal, .pro:
      return paidEntitlements
    }
  }

  /// Returns the numeric limit for the given quota at the given subscription tier.
  public static func limit(for tier: SubscriptionTier, quota: Quota) -> Int {
    switch (tier, quota) {
    case (.free, .watchZones):
      return 1
    case (.personal, .watchZones):
      return 3
    case (.pro, .watchZones):
      return Int.max
    }
  }

  /// Returns whether the given tier grants the specified entitlement.
  public static func hasEntitlement(_ entitlement: Entitlement, for tier: SubscriptionTier)
    -> Bool {
    entitlements(for: tier).contains(entitlement)
  }

  /// Returns whether the given tier can add another item given the current count for a quota.
  public static func canAdd(for tier: SubscriptionTier, currentCount: Int, quota: Quota) -> Bool {
    currentCount < limit(for: tier, quota: quota)
  }
}

File: mobile/android/domain/src/main/kotlin/uk/towncrierapp/domain/subscriptions/EntitlementMap.kt
package uk.towncrierapp.domain.subscriptions

/**
 * Maps subscription tiers to their granted entitlements and quota limits.
 * Must remain in sync with the API's `EntitlementMap.cs`. Port of iOS
 * `EntitlementMap` — a caseless object since it has no instances, only
 * static-style lookups.
 */
public object EntitlementMap {
    private val paidEntitlements: Set<Entitlement> =
        setOf(Entitlement.STATUS_CHANGE_ALERTS, Entitlement.DECISION_UPDATE_ALERTS, Entitlement.HOURLY_DIGEST_EMAILS)

    /** Returns the set of entitlements granted to the given subscription tier. */
    public fun entitlements(tier: SubscriptionTier): Set<Entitlement> =
        when (tier) {
            SubscriptionTier.FREE -> emptySet()
            SubscriptionTier.PERSONAL, SubscriptionTier.PRO -> paidEntitlements
        }

    /** Returns the numeric limit for the given quota at the given subscription tier. */
    public fun limit(
        tier: SubscriptionTier,
        quota: Quota,
    ): Int =
        when (quota) {
            Quota.WATCH_ZONES ->
                when (tier) {
                    SubscriptionTier.FREE -> 1
                    SubscriptionTier.PERSONAL -> 3
                    SubscriptionTier.PRO -> Int.MAX_VALUE
                }
        }

    /** Returns whether the given tier grants the specified entitlement. */
    public fun hasEntitlement(
        entitlement: Entitlement,
        tier: SubscriptionTier,
    ): Boolean = entitlements(tier).contains(entitlement)

    /** Returns whether the given tier can add another item given the current count for a quota. */
    public fun canAdd(
        tier: SubscriptionTier,
        currentCount: Int,
        quota: Quota,
    ): Boolean = currentCount < limit(tier, quota)
}
